Adding subword  realted parts (#20)
cadia-lvl/samromur-asr: s5_subwords/readme.md

# Subword recipe for Icelandic

> This is a subword-based recipe for an Icelandic ASR build using the Kaldi toolkit. Running the recipe on other languages only requires modifications to the dataset-specific data preparation steps. Subword-based ASR models have the advantage of being able to create out-of-vocabulary words. 

Reviewer: ```suggestion
This is a subword-based recipe for an Icelandic ASR built using the Kaldi toolkit. Running the recipe on other languages only requires modifications to the dataset-specific data preparation steps. Subword-based ASR models have the advantage of being able to transcribe out-of-vocabulary words. 
```

## Running the recipe 

> In the folder "run" are the scripts used to interface the recipe. The acoustic and subword modelling are in separate scripts. The subword related modifications are only done to the lexicon and language model, except that the phonetic units used for the acoustic model are graphemes instead of phonemes. 

Reviewer: ```suggestion
In the folder "run" are the scripts used to interface the recipe. The acoustic and subword modelling are in separate scripts. Most of the subword related modifications affect the lexicon and language model, except that the acoustic model uses graphemes instead of phoenemes for the phonetic units.

### run/run_am.sh

> Start by running the script "run/run_am.sh", it has all the steps related to the acoustic model training. It creates a standard LDA+MLLT+SAT system followed by a TDNN model, with i-vectors and speed perturbation. The script is a standard Kaldi recipe except for the creation of the dictionary as it uses the text corpus to create a grapheme based lexicon instead of a phoneme based model. The "lang" and "dict" used for the acoustic model training are prefixed with "base" to distinguish them from their subword counterparts, which will be created later. The "lang" and "dict" are derived from the text corpus, so no g2p translation is required. Note that the text corpus must be normalized; one sentence per line, all characters should be lower or upper case, and all non-character related signs like numbers or punctuations should be removed. 

Reviewer: ```suggestion
Start by running the script [run/run_am.sh](run/run_am.sh), it has all the steps related to the acoustic model training. It creates a standard LDA+MLLT+SAT system followed by a TDNN model, with i-vectors and speed perturbation. The script is a standard Kaldi recipe except for the creation of the dictionary as it uses the text corpus to create a grapheme based lexicon instead of a phoneme based model. The "lang" and "dict" used for the acoustic model training are prefixed with "base" to distinguish them from their subword counterparts, which will be created later. The "lang" and "dict" are derived from the text corpus, so no g2p translation is required. Note that the text corpus must be normalized; one sentence per line, all characters should be lower or upper case, and all non-character related signs like numbers or punctuation should be removed. 
```
lower or upper case? so it's not allowed to have both?

### Subword model

> The subword realted steps are in run/run_sw.sh. There are scripts for three subword tokenizers:

Reviewer: ```suggestion
The subword related steps are in [run/run_sw.sh](run/run_sw.sh). There are scripts for three subword tokenizers:
```

 * [Unigram](https://arxiv.org/abs/1804.10959) 
 * [Byte Pair Encoding](https://arxiv.org/abs/1508.07909)
 * [Morfessor](https://morfessor.readthedocs.io/en/latest/)
 

> Please read about the tokenizer in their respective repositories, they all perform the same task, but the WER results will differ for each method. The default tokenizer is a "bpe", all scripts for that tokenizer are provided with this repository. To run the tokenizers "unigram" and the "sp_bpe" [Google's Sentence Piece](https://github.com/google/sentencepiece) is needed, a code snippet for a python virtual environment is provided along with a requirments.txt file. 

Reviewer: 1. what do you mean by specific repositories? The links are only to their papers, except for the Morfessor one.
2. Can you add Google's Sentence Piece repo as a submodule?
```suggestion
Please read about the tokenizer in their respective repositories, they all perform the same task, but the WER results will differ for each method. Our default tokenizer is "bpe", all scripts for that tokenizer are provided with this repository. To run the tokenizers "unigram" and the "sp_bpe," [Google's Sentence Piece](https://github.com/google/sentencepiece) is needed, a code snippet for a python virtual environment is provided along with a requirements.txt file. 
```

> For each subword tokenization method, there are two parameters, vocabulary size and boundary marking style. The vocabulary size is a parameter that controls how aggressively the words are tokenized, lower values mean fewer and smaller subword units and vice-versa for higher values. The boundary marking styles are the following (the sentence "dog walks" is used as an example): 

Reviewer: dog walks is a phrase not a sentence

Reviewer: but maybe you mean sentence in the asr sense

* right-marked (r):  do+ g w+ alk+ s
* left-marked (l):  do +g w +alk +s
* left-right-marked (lr): do+ +g w+ +alk+ +s (The default) 
* word-boundary (wb): do g + w alk s 

The token used to mark the subword boundary can be any sign given that it's not a part of the language, in this setup the default is "+".

### run/run_sw.sh

> The first step in the script is to train the subword tokenization model. Depending on its size, the text corpus or a subset of it is used for that. We have been using 4 million lines of text which should be more than enough. When the tokenizer is trained that model is used to tokenize the entire corpus. This subword tokenized text corpus will be used to prepare the Language model. There are three different subword tokenizers in the script Unigram, 

Reviewer: Instead of "Depending on its size, the text corpus or a subset of it is used for that." You should put in what you had in the script itself: "If you have a corpus with more than 4 million sentences, create 4 million sentence subset for training." and mention that "This limit is hard coded into the configs for the SP models."

Reviewer: Also, you should number each of these steps in the run/run_sw.sh section like 1. 2. 3. 4.

> The next step is to create the dictionary folder which that has the lexicon and other files that Kaldi needs. They are all made by extracting information from the subword tokenized text corpus. 

Reviewer: ```suggestion
The second step is to create the dictionary folder which has the lexicon and other files that Kaldi needs. They are all made by extracting information from the subword tokenized text corpus. 
```

> The next step is to prepare the "lang" directory. For that, an external package is needed, which was developed by [Aalto University](https://github.com/aalto-speech/subword-kaldi.git). It is used to create the L.fst for the boundary marking style. Please read their paper, found in the repository, for more information. 

Reviewer: can you add this repo as a git submodule?

```suggestion
The next step is to prepare the "lang" directory. For that, an external package, [subword-kaldi](https://github.com/aalto-speech/subword-kaldi.git) is needed, which was developed by [Aalto University](https://github.com/aalto-speech/subword-kaldi.git). It is used to create the L.fst for the boundary marking style. Please read their paper, found in the repository, for more information. 
```

Author: Will do

The last step before decoding is to create the language model. The n-gram count has to be higher than in traditional language models. The default is 6-gram for the decoding LM and 8-gram for the rescoring LM.

### run/tdnn_decode.sh

> This step is the final step in "run/run_sw.sh". In this step, the decoding graph is compiled. Next, the correct "wer_output_filter" is created depending on the boundary marking style used. This filter is a set of sed commands that will remove the space and boundary marker between individual subword units. Finally the decoding is done for the decoding and rescoring LM.

Reviewer: ```suggestion
This step is the final step in [run/run_sw.sh](run/run_sw.sh). In this step, the decoding graph is compiled. Next, the correct "wer_output_filter" is created depending on the boundary marking style used. This filter is a set of sed commands that will remove the space and boundary marker between individual subword units. Finally the decoding is done for the decoding and rescoring LM.
```
